File: Dec2020/26Dec/LuckyNumbersInAMatrix.java
/*
Problem Name: 1380. Lucky Numbers in a Matrix
Problem Link: https://leetcode.com/problems/lucky-numbers-in-a-matrix/
*/
class Solution {
    public List<Integer> luckyNumbers (int[][] matrix) {
        int n = matrix.length, m = matrix[0].length;
        int row[] = new int[n], col[] = new int[m];
        for(int i=0;i<n;i++){
            int min = Integer.MAX_VALUE;
            for(int j=0;j<m;j++){
                min = Math.min(min,matrix[i][j]);
            }
            row[i] = min;
        }
        for(int i=0;i<m;i++){
            int max = Integer.MIN_VALUE;
            for(int j=0;j<n;j++){
                max = Math.max(max,matrix[j][i]);
            }
            col[i] = max;
        }
        List<Integer> lucky = new ArrayList<Integer>();
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(row[i] == col[j])
                    lucky.add(row[i]);
            }
        }
        return lucky;
    }
}

File: Dec2020/26Dec/LuckyNumbersInAMatrix.cpp
/*
Problem name: 1380. Lucky Numbers in a Matrix
Problem link: https://leetcode.com/problems/lucky-numbers-in-a-matrix/
*/

class Solution {
public:
    vector<int> luckyNumbers (vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<int> row_min(m,INT_MAX),col_max(n,INT_MIN);
        
        for(int i=0;i<m;i++)
            for(int j=0;j<n;j++)
                row_min[i] = min(row_min[i],matrix[i][j]);

        for(int i=0;i<n;i++)
            for(int j=0;j<m;j++)
                col_max[i] = max(col_max[i],matrix[j][i]);
        
        vector<int> res;
        for(int i=0;i<m;i++)
            for(int j=0;j<n;j++)
                if(row_min[i]==matrix[i][j] and col_max[j]==matrix[i][j])
                    res.push_back(matrix[i][j]);
        
        return res;
    }
};
